File: GO/Structural/adapter.go
package main

import "fmt"

func main() {
	client := &Client{}
	mac := &Mac{}

	client.InsertLightningConnectorIntoComputer(mac)
	fmt.Println("================================================================")

	windowsMachine := &Windows{}
	windowsMachineAdapter := &WindowsAdapter{
		windowsMachine: windowsMachine,
	}

	client.InsertLightningConnectorIntoComputer(windowsMachineAdapter)
}

type Client struct {
}

func (c *Client) InsertLightningConnectorIntoComputer(com Computer) {
	fmt.Println("Client inserts Lightning connector into computer.")
	com.InsertIntoLightingPort()
}

// Computer is a client interface
type Computer interface {
	InsertIntoLightingPort()
}

// Mac service
type Mac struct {
}

func (m *Mac) InsertIntoLightingPort() {
	fmt.Println("Lightning connector is plugged into mac machine.")
}

// Windows is an unknown service
type Windows struct {
}

func (w *Windows) insertIntoUSBPort() {
	fmt.Println("USB connector is plugged into windows machine.")
}

// WindowsAdapter is an adapter
type WindowsAdapter struct {
	windowsMachine *Windows
}

func (wa *WindowsAdapter) InsertIntoLightingPort() {
	fmt.Println("Adapter converts Lightning signal to USB.")
	wa.windowsMachine.insertIntoUSBPort()
}

File: DesignPatterns/python/Structural/adapter.py
from __future__ import annotations
from abc import ABC, abstractmethod

"""
Intent

Pattern that allows objects with incompatible interfaces to collaborate.
"""


class Computer(ABC):
    @abstractmethod
    def InsertIntoLightingPort(self):
        pass


class Client:
    def InsertLightningConnectorIntoComputer(self, computer: Computer):
        print("Client inserts Lightning connector into computer.")
        computer.InsertIntoLightingPort()


class Mac(Computer):
    def InsertIntoLightingPort(self):
        print("Lightning connector is plugged into mac machine.")


class Windows:
    def InsertIntoUSBPort(self):
        print("USB connector is plugged into windows machine.")


class WindowsAdapter(Computer):
    def __init__(self, windowsMachine: Windows):
        self._windowsMachine = windowsMachine

    def InsertIntoLightingPort(self):
        print("Adapter converts Lightning signal to USB.")
        self._windowsMachine.InsertIntoUSBPort()


if __name__ == '__main__':
    client = Client()

    mac = Mac()
    client.InsertLightningConnectorIntoComputer(mac)
    print("================================================================")

    windowsMachine = Windows()
    windowsMachineAdapter = WindowsAdapter(windowsMachine)
    client.InsertLightningConnectorIntoComputer(windowsMachineAdapter)

